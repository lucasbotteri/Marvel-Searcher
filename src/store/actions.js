import {
  SET_LOADING_CHARACTERS_LIST,
  SET_CHARACTERS_LIST,
  SET_CHARACTER_SELECTED,
  HIDE_MODAL,
  SET_COMICS,
  SET_LOADING_COMICS,
  SET_COMIC,
  SET_LOADING_COMIC,
  SAVE_FAV,
  SHOW_FAVS,
  REMOVE_FAV,
  HIDE_FAVS,
} from "./constants";
import {
  getCharacterByName,
  getRandomCharacter,
  getCharacterComics,
  getComic,
} from "../service";

export const setLoadingCharacters = () => ({
  type: SET_LOADING_CHARACTERS_LIST,
});

export const setCharacters = (characters) => ({
  type: SET_CHARACTERS_LIST,
  payload: characters,
});

export const setCharacterSelected = (id, name) => ({
  type: SET_CHARACTER_SELECTED,
  payload: { id, name },
});

export const hideModal = () => ({
  type: HIDE_MODAL,
});

export const setComics = (comics) => ({
  type: SET_COMICS,
  payload: comics,
});

export const setLoadingComics = () => ({
  type: SET_LOADING_COMICS,
});

export const setComic = (comics) => ({
  type: SET_COMIC,
  payload: comics,
});

export const setLoadingComic = () => ({
  type: SET_LOADING_COMIC,
});

export const saveFav = (character) => ({
  type: SAVE_FAV,
  payload: character,
});

export const showFavs = () => ({
  type: SHOW_FAVS,
});

export const hideFavs = () => ({
  type: HIDE_FAVS,
});

export const removeFav = (id) => ({
  type: REMOVE_FAV,
  payload: { id },
});

export const searchCharacter = async (dispatch, characterName) => {
  dispatch(setLoadingCharacters());
  const characters = await getCharacterByName(characterName);
  dispatch(setCharacters(characters));
};

export const searchRandomCharacter = async (dispatch) => {
  dispatch(setLoadingCharacters());
  const characters = await getRandomCharacter();
  dispatch(setCharacters(characters));
};

export const searchCharacterComics = async (
  dispatch,
  characterId,
  comicFilter
) => {
  dispatch(setLoadingComics());
  const comics = await getCharacterComics(characterId, comicFilter);
  dispatch(setComics(comics));
};

export const searchComic = async (dispatch, comicId) => {
  dispatch(setLoadingComic());
  const comic = await getComic(comicId);
  dispatch(setComic(comic));
};
