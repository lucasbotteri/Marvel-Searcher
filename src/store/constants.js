export const LOADING_CHARACTERS_LIST = "LOADING_CHARACTERS_LIST";
export const SET_CHARACTERS_LIST = "SET_CHARACTERS_LIST";
export const SET_CHARACTER_SELECTED = "SET_CHARACTER_SELECTED";
export const HIDE_MODAL = "HIDE_MODAL";
export const SET_COMICS = "SET_COMICS";
export const SET_LOADING_COMICS = "SET_LOADING_COMICS";
