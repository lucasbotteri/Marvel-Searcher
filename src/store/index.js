export { default as StoreProvider, StoreContext } from "./StoreProvider";
