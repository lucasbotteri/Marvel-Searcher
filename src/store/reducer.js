import {
  LOADING_CHARACTERS_LIST,
  SET_CHARACTERS_LIST,
  SET_CHARACTER_SELECTED,
  HIDE_MODAL,
} from "./constants";

const reducer = (state, action) => {
  switch (action.type) {
    case LOADING_CHARACTERS_LIST:
      return { ...state, characters: [], isCharacterListLoading: true };
    case SET_CHARACTERS_LIST:
      return {
        ...state,
        characters: action.payload,
        isCharacterListLoading: false,
        isShowingModal: false,
      };
    case SET_CHARACTER_SELECTED:
      return {
        ...state,
        characterSelected: action.payload,
        isShowingModal: true,
      };
    case HIDE_MODAL:
      return {
        ...state,
        isShowingModal: false,
      };
    default:
      throw new Error("Invalid Action");
  }
};

export default reducer;
