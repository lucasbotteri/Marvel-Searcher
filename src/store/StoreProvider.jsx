import React, { useReducer } from "react";
import PropTypes from "prop-types";
import reducer from "./reducer";
import * as actions from "./actions";

const INITAL_STATE = {
  characters: [],
  loadingCharacters: false,
  fetchedCharactersSuccessful: true,
};

export const StoreContext = React.createContext();

const StoreProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, INITAL_STATE);
  const value = {
    ...state,
    fetchCharacters: (data) => actions.fetchCharacters(dispatch, data),
  };
  return (
    <StoreContext.Provider value={value}>{children}</StoreContext.Provider>
  );
};

StoreProvider.propTypes = {
  children: PropTypes.element.isRequired,
};

export default StoreProvider;
