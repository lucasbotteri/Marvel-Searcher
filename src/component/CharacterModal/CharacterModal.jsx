import React, { useContext } from "react";
import Modal from "../Modal";
import { StoreContext } from "../../store";
import ComicList from "../ComicList";

const CharacterModal = () => {
  const {
    isShowingModal,
    characterSelected,
    comics,
    hideModal,
    isLoadingComics,
  } = useContext(StoreContext);
  return (
    <Modal
      isShowing={isShowingModal}
      title={characterSelected.name}
      onClose={hideModal}
    >
      {isLoadingComics ? "Loading comics..." : <ComicList comics={comics} />}
    </Modal>
  );
};

export default CharacterModal;
