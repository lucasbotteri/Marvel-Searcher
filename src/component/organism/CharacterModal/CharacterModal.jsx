import React, { useContext } from "react";
import Modal from "../../molecule/Modal";
import { StoreContext } from "../../../store";
import ComicList from "../ComicList";

const CharacterModal = () => {
  const {
    isShowingModal,
    characterSelected,
    comics,
    hideModal,
    isLoadingComics,
  } = useContext(StoreContext);
  return isShowingModal ? (
    <Modal
      isShowing={isShowingModal}
      title={characterSelected.name}
      onClose={hideModal}
    >
      {isLoadingComics ? "Loading comics..." : <ComicList comics={comics} />}
    </Modal>
  ) : null;
};

export default CharacterModal;
