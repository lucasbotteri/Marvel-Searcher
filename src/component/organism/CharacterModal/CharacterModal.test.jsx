import React from "react";
import { mount } from "enzyme";
import CharacterModal from "./CharacterModal";
import { StoreContext } from "../../../store";

const contextStore = {
  isShowingModal: true,
  characterSelected: { id: 1014997, name: "spiderman" },
  comics: [],
  hideModal: () => {},
  isLoadingComics: false,
};

const mountCharacterModal = (overrideStore) => {
  return mount(
    <StoreContext.Provider value={{ ...contextStore, ...overrideStore }}>
      <CharacterModal />
    </StoreContext.Provider>
  );
};

describe("CharacterModal", () => {
  it("Should render without crashing", () => {
    const wrapper = mountCharacterModal();
    expect(wrapper).toBeTruthy();
  });

  it("Should render loading", () => {
    const wrapper = mountCharacterModal({ isLoadingComics: true });
    expect(wrapper.containsMatchingElement("Loading comics...")).toBeTruthy();
  });

  it("Should render a comic list", () => {
    const wrapper = mountCharacterModal();
    expect(wrapper).toBeTruthy();
  });

  it("Should not render modal", () => {
    const wrapper = mountCharacterModal({ isShowingModal: false });
    expect(wrapper).toMatchObject({});
  });

  it("Should match snapshot", () => {
    const wrapper = mountCharacterModal();
    expect(wrapper).toMatchSnapshot();
  });
});
