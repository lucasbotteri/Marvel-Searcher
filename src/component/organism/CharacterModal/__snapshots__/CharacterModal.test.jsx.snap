// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterModal Should match snapshot 1`] = `
<CharacterModal>
  <Modal
    onClose={[Function]}
    title="spiderman"
  >
    <styled.div>
      <div
        className="sc-AxjAm jLfDnH"
      >
        <styled.aside
          aria-labelledby="modal-title"
          role="dialog"
        >
          <aside
            aria-labelledby="modal-title"
            className="sc-AxirZ gEfLNE"
            role="dialog"
          >
            <Styled(FontAwesomeIcon)
              aria-label="close modal"
              icon={
                Object {
                  "icon": Array [
                    352,
                    512,
                    Array [],
                    "f00d",
                    "M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z",
                  ],
                  "iconName": "times",
                  "prefix": "fas",
                }
              }
              onClick={[Function]}
            >
              <FontAwesomeIcon
                aria-label="close modal"
                border={false}
                className="sc-AxhCb eTlDTE"
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      352,
                      512,
                      Array [],
                      "f00d",
                      "M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z",
                    ],
                    "iconName": "times",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                listItem={false}
                mask={null}
                onClick={[Function]}
                pull={null}
                pulse={false}
                rotation={null}
                size={null}
                spin={false}
                swapOpacity={false}
                symbol={false}
                title=""
                transform={null}
              >
                <svg
                  aria-hidden="true"
                  aria-label="close modal"
                  className="svg-inline--fa fa-times fa-w-11 sc-AxhCb eTlDTE"
                  data-icon="times"
                  data-prefix="fas"
                  focusable="false"
                  onClick={[Function]}
                  role="img"
                  style={Object {}}
                  viewBox="0 0 352 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </FontAwesomeIcon>
            </Styled(FontAwesomeIcon)>
            <styled.h3
              id="modal-title"
            >
              <h3
                className="sc-AxhUy hSmtMl"
                id="modal-title"
              >
                 
                spiderman
              </h3>
            </styled.h3>
            <styled.div>
              <div
                className="sc-AxiKw bXxfRn"
              >
                <ComicList
                  comics={Array []}
                >
                  <styled.ul>
                    <ul
                      className="sc-fzplWN BkLwV"
                    />
                  </styled.ul>
                </ComicList>
              </div>
            </styled.div>
          </aside>
        </styled.aside>
      </div>
    </styled.div>
  </Modal>
</CharacterModal>
`;
